common/channel/src/libra_channel.rs

```diff
@@ -0,0 +1,151 @@
+            max_queue_size,
+        }
+    }
+
+    fn add_message(&mut self, message: T) {
+        let validator = message.get_validator();
+        if !self.per_validator_queue.contains_key(validator) {
+            self.per_validator_queue.insert(
+                validator.to_string(),
+                VecDeque::with_capacity(self.max_queue_size + 1),
+            );
+        }
+        if let Some(q) = self.per_validator_queue.get_mut(validator) {
+            q.push_back(message);
+            if q.len() == self.max_queue_size + 1 {
+                match self.queue_type {
+                    // Drop old message for FIFO
+                    Type::FIFO => q.pop_front(),
+                    // Drop new message for LIFO
+                    Type::LIFO => q.pop_back(),
+                };
+            }
+        }
+    }
+
+    // TODO: Come up with a better data structure to efficiently
+    // do a round robin among validators
+    fn get_message(&mut self) -> Option<T> {
+        for q in self.per_validator_queue.values_mut() {
+            if !q.is_empty() {
+                return match self.queue_type {
+                    Type::FIFO => q.pop_back(),
+                    Type::LIFO => q.pop_front(),
+                };
+            }
+        }
+        None
+    }
+}
+
+struct SharedState<T> {
+    /// The data stored by this Channel
+    /// TODO This needs to be updated to a data structure which can do round robin for validators
+    data: PerValidatorQueue<T>,
+
+    /// The thread can use this after calling `put` to tell
+    /// `Receiver`'s task to wake up, and process the next item
+    waker: Option<Waker>,
+}
+
+/// The sending end of the Channel
+#[derive(Clone)]
+pub struct Sender<T: ValidatorMessage> {
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+/// The receiving end of the Channel
+pub struct Receiver<T: ValidatorMessage> {
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+impl<T: ValidatorMessage> Sender<T> {
+    /// TODO: We can have this return the old data if the new data is replacing some existing data
+    /// This puts the data into the internal data structure. This is a non-blocking synchronous call
+    pub fn put(&mut self, data: T) {
+        let mut shared_state = self.shared_state.lock().unwrap();
+        // TODO: Enforce maximum size
+        shared_state.data.add_message(data);
+        if let Some(w) = shared_state.waker.take() {
+            w.wake();
+        }
+    }
+}
+
+impl<T: ValidatorMessage> Stream for Receiver<T> {
+    type Item = T;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut shared_state = self.shared_state.lock().unwrap();
+        if let Some(val) = shared_state.data.get_message() {
+            Poll::Ready(Some(val))
+        } else {
+            shared_state.waker = Some(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+}
+
+impl<T: ValidatorMessage> FusedStream for Receiver<T> {
```

Comment:
Could you elaborate on this one a little bit more? Why is it needed and why is it always false? :)

Reply:
Sorry I missed this. The `FusedStream` is a trait extension for the `Stream` trait.  

`Stream` trait's specification has undefined behaviour when calling `poll_next()` after a Stream has ended (by sending a `Poll::Ready(None)` in its last response). 

`FusedStream` removes this undefined behaviour by providing a `is_terminated` function to allow a caller (The `select!()` macro in this case) to decide whether to call `poll_next()` on the underlying `Stream`.

`is_terminated` may also return true if a `Stream` has become inactive and can no longer make progress and should be ignored or dropped rather than being polled again.

In the case of LibraChannel, I don't see this channel ending anytime, so I have set `is_terminated` to false. 

If we want to provide a provision to close the channel from the sender side, we can do so by providing a `close_channel()` method on the Sender struct which sets a boolean value in the `SharedState` struct, but I don't see a case for that now.